### DISCUSSION TITLE
General questions regarding the ICP algorithm

### DISCUSSION THREAD

**Discussion by jkissi** [marcomusy/vedo#507] [Q&A] [answered]
Hi Marco 
I need to do some cloud registration stuff and have a few questions about your implementation of Iterative Closest Point:

1. Can I output the transformation matrix as a numpy array? 
2. Is it possible to out the matrix at various stages of the iteration? Say the initial and then the final, for example?
3. Are we able to access the **inlier** correspondences? for calculations such as the **fitness** of the alignment and the **inlier rmse**?
4. Which implementation of ICP are you using? Is there the possibility to use point-to-point or point-to-plane or do you only have one version?

Sorry for the quick-fire questions, hope you don't mind too much! 
Cheers!
J

**Comment by marcomusy** [comment, marked as answer]
> 1. Can I output the transformation matrix as a numpy array?

Yes, well you can save it to file with `vedo.io.writeTransform(some_mesh.transform)`, if you need it as numpy array you may need to copypaste that function and adapt it (it's easy).


> 2. Is it possible to out the matrix at various stages of the iteration? Say the initial and then the final, for example?

No, I'm afraid that is not possible. Only the final iteration is output.

> 3. Are we able to access the **inlier** correspondences? for calculations such as the **fitness** of the alignment and the **inlier rmse**?

it's a good question, it seems that is not passed in the upstream vtk. I must investigate more about it... One thing you can do is to compute the distance of one pointcloud to another pointcloud or Mesh with `pc1.distanceToMesh(pc2)` , see #495 

> 4. Which implementation of ICP are you using? Is there the possibility to use point-to-point or point-to-plane or do you only have one version?

yes - there are various linear and non linear registration methods. Check out the examples in `basic/align*.py` and `examples/advanced/warp*.py`

**Reply by jkissi** [reply]
Hi Marco
As always thanks for you reply. Its actually important for me to get some kind of definable measure of the tightness of the fit between the point clouds I am measuring and for that I need information about which points make up the inlier correspondences and their distances. 

Do you think there is some way that I can [reliably] approximate this information after the fact? I noticed in one of the basic examples you have a function called closestPointTo()....could I incorporate that somehow?

J

**Reply by jkissi** [reply]
Hi 
Another quick question: Can you show how to convert an array of points to a Mesh? I know it sounds simple but I an trying to use the distanceToMesh() function and I keep getting the following error:

![distance_mesh_edit](https://user-images.githubusercontent.com/8246894/140413591-30c54db6-19a2-4361-84bc-f52c27ecdac9.png)

Can you let me know where I am going wong? Or is there a way to use the distanceToMesh() function with a Points object..? 

J

**Reply by marcomusy** [reply]
uhm this is actually a problem because I always thought that it would work with pointclouds too.... So I have to look into it ...

i'll convert this into an Issue to better keep track of it.

**Reply by jkissi** [reply]
Oh ok, cheers! Any ideas about my earlier question? Just wondering if you have any suggestions, cheers.

**Reply by marcomusy** [reply]
> Oh ok, cheers! Any ideas about my earlier question? Just wondering if you have any suggestions, cheers.

yes - you can use `closestPoint()` in the meanwhile to assess how well aligned are the two clouds (that should be relatively fast because there is an OBBTree operating behind the scene). You can check on subset of the points if it's too slow...
